Add Credentials chain

d2a4u opened the pull request:
Main changes are in:
- native/src/main/scala/sqs4s/auth/CredProvider.scala
- native/src/main/scala/sqs4s/api/lo/Action.scala
- native/src/main/scala/sqs4s/api/lo/SignedRequest.scala
- native/src/test/scala/sqs4s/auth/CredProviderSpec.scala

The rest of changes are just moving things around to avoid major release version bump. I still need to write more Unit tests and think about away to integration test this if possible.

d2a4u commented on native/src/main/scala/sqs4s/auth/CredProvider.scala:
```diff
@@ -0,0 +1,159 @@
+    Ref.of[F, Option[Credential]](None).flatMap { cred =>
+      val provider =
+        new InstanceProfileCredProvider[F](client, cred, ttl, refreshBefore)
+      provider.refreshAsync.compile.drain.start.as(provider)
```
Still not 100% happy with this line here. Basically what it does it run a fs2 Stream in a `Fiber` so it is a bit like fire and forget. What that stream does is to run `refresh` to rotate the cred periodically.

BalmungSan replied:
Do we really need that? Isn't it easier _(even better)_ to just run `refresh` if we get an **Unauthorized**?
If after the `refresh` we get another **Unauthorized** then fail with an error.

d2a4u replied:
that is what I'm already doing [here](https://github.com/d2a4u/sqs4s/pull/128/files#diff-3b5fb231906d76041d851d15b053f5f1R40). There are 2 ways to use this `InstanceProfileCredProvider`, 1st one is like you described, which already covered, 2nd one is using this `asyncRefresh`, again, this is something that the Java SDK has and I was thinking it is actually pretty cool feature to have, to reduce the number of 401 error and retries. Maybe I should make `refresh` returns `Unit` instead.

BalmungSan replied:
> that is what I'm already doing here.

Yeah, I just read my comment again and I realized that it gave the impression that I was saying you didn't do that, sorry.
My intention was to say that should be enough.

> I was thinking it is actually pretty cool feature to have, to reduce the number of 401 error and retries.

I also think it is a cool feature.
However, I believe that one makes more sense for other services that require HA and low throughput.
SQS is considered for simple applications were a delay shouldn't be a problem _(especially considering we have to pull over some rate instead of getting a push)_. So, my humble opinion is that the retry on failure should be more than enough.

_Neverthless, just because I as you want to make cool software, take a look to [this](https://www.reddit.com/r/scala/comments/ho6hr8/fp_approach_to_manage_the_lifecycle_of_an_api/fxgp6j9?utm_source=share&utm_medium=web2x) basically the idea would be that the `async` one returns a resource which under the hood uses a **Signal**. Also, it would be good if the `async` one overrides `refresh` as a no-op_.

> Maybe I should make refresh returns Unit instead.

+1

BalmungSan commented on native/src/main/scala/sqs4s/auth/CredProvider.scala:
```diff
@@ -0,0 +1,159 @@
+) extends TemporaryCredProvider[F]
+    with Http4sClientDsl[F] {
+
+  private val linkLocal = uri"http://169.254.169.254"
```
What is this address for?

d2a4u replied:
it is the address to retrieve cred from per doc [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials) and [AWS SDK](https://github.com/aws/aws-sdk-java/blob/542606ddd5ce6f0c3570ef7488c4f5479c6425bb/aws-java-sdk-core/src/main/java/com/amazonaws/util/EC2MetadataUtils.java#L73)

BalmungSan replied:
Oh, I didn't noticed before they were using a static IP instead of some DNS, that suprises me.

I would:

1. Rename it as `awsLocalLink`, mostly because the name _local_ made me think it was for testing and that you forgot to remove it.
2. Add your response as a comment in the code, since I am sure any other developer reading the code would want an explanation of why having a hardcoded IP address.

BalmungSan commented on native/src/main/scala/sqs4s/auth/CredProvider.scala:
```diff
@@ -0,0 +1,159 @@
+  def secretKey: F[String] = secretK.pure[F]
+}
+
+class InstanceProfileCredProvider[F[_]: Sync: Clock: Timer](
```
I would call this `RoleCredProvider` since this should work also in **ECS**, **EKS**, and even **Lambda**.

Also, in the future, we could add an `EnvCredProvider` which looks for the access tokens in the environment and another one `ConfigFileCredProvider` that searches for the access tokens in  the `.aws/credentials` file.

d2a4u replied:
the naming is to mirror what AWS SDK provide: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/InstanceProfileCredentialsProvider.html. My plan is to provider Env and Java properties next, and have a top level one to look for all available options just like the SDK.

BalmungSan replied:
> the naming is to mirror what AWS SDK provide:

Ok makes sense. I guess they named like that from a beginning and couldn't change it later, but that is just speculation.

> have a top level one to look for all available options just like the SDK.


Amazing!

d2a4u commented:
@BalmungSan I have rework this completely, sorry 😞. I think the `Provider` pattern is a bit Java-ish and we don't need to stick to it because we have cats effect `Resource` and it is already representing a `Provider`. However, I cannot test this atm since I haven't set up anything other than SQS in my personal AWS account (there are some unit tests though). It would take a while for me to set up EC2 and stuff to test.

BalmungSan commented:
@d2a4u 

> I have rework this completely, sorry

No problem.

> I think the Provider pattern is a bit Java-ish and we don't need to stick to it because we have cats effect Resource and it is already representing a Provider.

Agree.

> However, I cannot test this atm since I haven't set up anything other than SQS in my personal AWS account 

I will try to test it ASAP.

---

However, I believe the current implementation doesn't work. Yeah, the refresh will be running in the background, but the values of the credentials won't be modified.

So, once you do `credentialsR.use { c => foo(c) }` the value of `c` would be set and it is immutable, so if they expire during the execution of `foo` one would get an error. Of course, if I reuse the resource, I would get a new set of credentials, but I believe that is not what we want.

Did you checkout the approach suggested by Paul I liked before? I believe that is the best way to do this.
Either returning a **Resource[F, Signal[F, Credentials]]** so the user is the one that has to call `get` on the signal.
Or returning a **Resource[F, Credentials]** where the methods to access the credentials info aren't pure but rather effectual, and are implemented using the **Signal**.

d2a4u commented:
> However, I believe the current implementation doesn't work. Yeah, the refresh will be running in the background, but the values of the credentials won't be modified.
> 
> So, once you do `credentialsR.use { c => foo(c) }` the value of `c` would be set and it is immutable, so if they expire during the execution of `foo` one would get an error. Of course, if I reuse the resource, I would get a new set of credentials, but I believe that is not what we want.

Yep, I'm an idiot, fix one thing, broken the other 😆. Should be fixed now.

> Did you checkout the approach suggested by Paul I liked before? I believe that is the best way to do this.
> Either returning a **Resource[F, Signal[F, Credentials]]** so the user is the one that has to call `get` on the signal.
> Or returning a **Resource[F, Credentials]** where the methods to access the credentials info aren't pure but rather effectual, and are implemented using the **Signal**.

Yeah, I looked into it and it is very similar to my approach. The only different is `Signal`. I don't want to expose this to the end user and also, `Signal` is backed by `Ref` anyway. I don't think we need `discrete` so I keep it simple by just using `Deferred` and `Ref`.

BalmungSan commented:
@d2a4u 

I just realized that currently, we had a problem.
When you do this:

```scala
accessKey <- credential.accessKey
secretKey <- credential.secretKey
```

Or this:

```scala
st <- cred.sessionToken
ak <- cred.accessKey
```

There is no guarantee that the two values are related.
There is a _(very small but still)_ possibility that between the first and the second call, the refresh was executed.

So I am thinking that the credentials should be plain case classes, but the configs should have an `F[Credential]` instead of a `Credential[F]`.

----

Also, I am still thinking that instead of having a **Stream** running in a fiber and doing all the back and forth with `Ref`.
I would be simpler to do something like:

```scala
Stream
  .repeatEval(refresh)
  .metered(ttl - refreshBefore)
  .holdResource
  .map { signal  =>
    signal.get.map { credentialResponse =>
      // Create a new instance of Credential using credentialResponse
    }
  }
```

Which returns a **Resource[F, F[Credential]]**.
_(and probably does the same under the hood, but it feels cleaner)_

d2a4u commented:
>There is no guarantee that the two values are related.
Good point. I have introduce a topic level `Credentials[F]` so it should be fixed now.
>Also, I am still thinking that instead of having a Stream running in a fiber and doing all the back and forth with Ref.
I managed to simplify this a lot so it looks better now. I have tried to use `Signal` and there isn't much difference to `Ref` impl, however, the problem is the way `chainResource` is implemented by chaining `F[..]` with `handleErrorWith`. So I need the underlying `F` which comes before `Resource` if that makes sense.

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -0,0 +1,250 @@
+  def envVarResource[F[_]: Sync]: Resource[F, Credentials[F]] =
+    Resource.liftF(envVar[F])
+
+  private def envVar[F[_]: Sync]: F[Credentials[F]] =
```
I believe this is one _(as well as the `sysProp` one)_ are wrong.
They will attempt to read the env variables or system properties in each use, whereas they should read them at initialization and then cache the results _(IMHO)_; this change is essential for my previous suggestion.

d2a4u commented:
@BalmungSan thank you so much for all of the comments and review effort, they are great and I have also learnt a lot. I have addressed all of the comments. In the meantime, I'm also setting up a dummy service to test this out in my AWS account.

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -59,23 +57,22 @@ object Credentials {
-    }
-  }
+  ): Resource[F, Credentials[F]] =
+    envVar[F] orElse sysProp[F] orElse instanceMetadata[F](
+      client,
+      ttl,
+      refreshBefore
+    )
```
I forgot to mention before that if all fail, we would get the error message of the last one.
So, in this case, it would be better to do something like this:

```scala
(
  envVar[F] orElse sysProp[F] orElse instanceMetadata[F](
    client,
    ttl,
    refreshBefore
  )
).handleErrorWith(_ => Sync[F].raiseError[Credential](NoValidAuthMethodError))
```

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -95,73 +92,73 @@ object Credentials {
+        )
+        val optSessionToken = Option(System.getenv(AWS_SESSION_TOKEN_ENV_VAR))
+
+        instance[F] {
+          (optAccessKey, optSecretKey, optSessionToken) match {
+            case (Some(accessKey), Some(secretKey), None) =>
+              BasicCredential(accessKey, secretKey)
+                .pure[F].widen
+
+            case (Some(accessKey), Some(secretKey), Some(token)) =>
+              TemporarySecurityCredential(
+                accessKey,
+                secretKey,
+                token
+              ).pure[F].widen
+
+            case _ =>
+              Sync[F].raiseError[Credential](NoEnvironmentVariablesFound)
+          }
```
So we still have the problem that the checking for the env vars is happening during the credentials access and not creation.
This means that if we use `chain` it will always use this one because its creation will never fail.

We can do this instead:

```scala
Sync[F].defer {
  // ...

  (optAccessKey, optSecretKey, optSessionToken) match {
    case (Some(accessKey), Some(secretKey), None) =>
      Credentials.basic[F](accessKey, secretKey).pure[F]

    case (Some(accessKey), Some(secretKey), Some(token)) =>
      Credentials.temporal[F](accessKey, secretKey, token).pure[F]

    case _ =>
      Sync[F].raiseError[Credential](NoEnvironmentVariablesFound)
  }
}
```

Where `Credentials.temporal` is implemented like `Credentials.basic`.

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -95,73 +92,73 @@ object Credentials {
+        val optSessionToken =
+          Option(System.getProperty(SESSION_TOKEN_SYSTEM_PROPERTY))
+
+        instance[F] {
+          (optAccessKey, optSecretKey, optSessionToken) match {
+            case (Some(accessKey), Some(secretKey), None) =>
+              BasicCredential(accessKey, secretKey)
+                .pure[F].widen
+
+            case (Some(accessKey), Some(secretKey), Some(token)) =>
+              TemporarySecurityCredential(
+                accessKey,
+                secretKey,
+                token
+              ).pure[F].widen
+
+            case _ =>
+              Sync[F].raiseError[Credential](NoSystemPropertiesFound)
+          }
```
Ditto.

d2a4u commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -0,0 +1,234 @@
+        Stream
+          .repeatEval(refresh)
+          .metered(ttl - refreshBefore)
+          .holdResource(init).map { sig =>
```
@BalmungSan I just realised with the `holdResource` approach, it is not possible to check if the Credential has expired and do a final refresh since we can only do `.get` on `Signal`. Any suggestion?

d2a4u replied:
this is what I used to have with `Ref` and `Fiber`:
```
Resource {
        fiberF.map { fiber =>
          val tmp = for {
            resp <- ref.get
            now <- Clock[F].realTime(TimeUnit.MILLISECONDS)
            updated <- if (resp.expiration.toEpochMilli < now) {
              refresh.flatMap(r => ref.updateAndGet(_ => r))
            } else {
              resp.pure[F]
            }
          } yield {
            TemporarySecurityCredential(
              updated.accessKeyId,
              updated.secretAccessKey,
              updated.token
            )
          }
          instance(tmp.widen) -> fiber.cancel
        }
      }
```

BalmungSan replied:
But do we really need that? I mean the signal should be always updated with some valid credentials.
If the credentials expired before the refresh is because your configuration is wrong, which for me should kill the application. What do you think?

d2a4u replied:
I think that is fair. I will stick with `Signal` 👍

d2a4u commented:
@BalmungSan I have added a feature to retrieve cred from metadata service within container and tested it manually with ECS and Fargate. However, I have not test this with EC2 since I'm not familiar with it. I tempted to merge this PR for now. WDYT?

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -68,101 +78,122 @@ object Credentials {
+      client,
+      ttl,
+      refreshBefore
+    ).handleErrorWith(_ =>
```
Technically speaking, this is handling the error of the last one and not of all of them.
The end result should be the same, but I would put the extra parenthesis around all the expression.

I would even do it like this:

```scala
def chain[F[_]: Concurrent: Clock: Timer](
  client: Client[F],
  ttl: FiniteDuration = 6.hours,
  refreshBefore: FiniteDuration = 5.minutes
): Resource[F, Credentials[F]] = {
  val env = envVar[F]
  val sys = sysProp[F]
  val container = containerMetadata[F](client, ttl, refreshBefore)
  val instance = instanceMetadata[F](client, ttl, refreshBefore)

  val tryAllInOrder = env orElse sys orElse container orElse instance

  tryAllInOrder.handleErrorWith { _ =>
    Resource.liftF(Sync[F].raiseError[Credentials[F]](NoValidAuthMethodError))
  }
}
```

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -68,101 +78,122 @@ object Credentials {
+        secretKey <- secretKeyF
+        optSessionToken <- optSessionTokenF
+      } yield {
+        optSessionToken.map(token =>
+          Credentials.temporal[F](
+            accessKey,
+            secretKey,
+            token
+          )).getOrElse(
+          Credentials.basic[F](
+            accessKey,
+            secretKey
+          )
+        )
```
`map` + `getOrEsle` = `fold`

```scala
optSessionToken.fold(
  ifEmpty = Credentials.basic[F](accessKey, secretKey)
) { token =>
  Credentials.temporal[F](
    accessKey,
    secretKey,
    token
  )
}
```

BalmungSan commented on native/src/main/scala/sqs4s/auth/Credentials.scala:
```diff
@@ -68,101 +78,122 @@ object Credentials {
+        secretKey <- secretKeyF
+        optSessionToken <- optSessionTokenF
+      } yield {
+        optSessionToken.map(token =>
+          Credentials.temporal[F](
+            accessKey,
+            secretKey,
+            token
+          )).getOrElse(
+          Credentials.basic[F](
+            accessKey,
+            secretKey
+          )
+        )
```
Ditto.

BalmungSan commented on native/src/main/scala/sqs4s/auth/SystemF.scala:
```diff
@@ -0,0 +1,24 @@
+
+object SystemF {
+  def env[F[_]: Sync](name: String): F[String] =
+    Sync[F].fromOption(
+      Option(System.getenv(name)),
+      NoEnvironmentVariablesFound(name)
+    )
```
This isn't suspending the side effect of reading from the environment.
It should be:

```scala
envOpt[F](name).flatMap { opt =>
  Sync[F].fromOption(opt, NoEnvironmentVariablesFound(name))
}
```

BalmungSan commented on native/src/main/scala/sqs4s/auth/SystemF.scala:
```diff
@@ -0,0 +1,24 @@
+  def envOpt[F[_]: Sync](name: String): F[Option[String]] =
+    Sync[F].delay(Option(System.getenv(name)))
+
+  def prop[F[_]: Sync](name: String): F[String] =
+    Sync[F].fromOption(
+      Option(System.getProperty(name)),
+      NoSystemPropertiesFound(name)
+    )
```
Ditto.
